### ISSUE TITLE
tablesorter disabled 

### ISSUE THREAD

**Issue by jvanasco** [Pylons/pyramid_debugtoolbar#148]
The SqlAlchemy panel didn't have tablesorting, so I looked into adding it.

When I looked at the code, saw this pull request: https://github.com/Pylons/pyramid_debugtoolbar/pull/63

Then I noticed that `pDebugSortable` was still on the table clqass

And finally i noticed that the tablesorter JS is loaded on the debugtoolbar, but this line is commented out at the end of the javascript :

// $(".pDebugSortable").tablesorter();

Is there any reason why this was disabled?  I couldn't find anything in the docs.  

re-enabling it on the console makes it work, however the table-headers are not decorated with icons ( ie, it works if you know to click the table heads )

**Comment by davidjb** [comment]
+1 for this.  Being able to sort the performance panel on its column headings is a must.  Uncommenting that line above enables things again.

**Comment by hamx0r** [comment]
+1 too.  Maybe add in Bootstrap Table for easy (and plentiful) search/sort options.  
http://wenzhixin.net.cn/p/bootstrap-table

**Comment by stevepiercy** [comment]
For those who have uncommented this:

```javascript
// $(".pDebugSortable").tablesorter();
```

...have you noticed any negative side effects or javascript errors?

If not, have any of you submitted a PR for this issue?

I'm reasonably sure that a PR with this change would get merged by the core contributors.

Note: +1's are unnecessary. The Issue tracker is not a voting platform.

**Comment by jvanasco** [comment]
I wouldn't just re-enable:

1. I don't believe many tables have that CSS class anymore (I recently stripped the even/odd CSS classes and don't remember that). A better selector would be all tables within the panel

2. A newer library/version would make sense.  A lot of table sorters were very slow.  Many js plugins are better now.

**Comment by hamx0r** [comment]
All i know is the uncommenting it in the master branch worked like a charm.  All tables in the Performance page became sortable and it was not slow.

**Comment by jvanasco** [comment]
At this point, I'm against just re-enabling it as-is.  Why?  This is broken and has been for 3 years(!).  Just uncommenting it will be building in equity/debt into something that may be wrong or of limited use (in terms of time).

Only 2 panels have the `pDebugSortable` class right now: Performance and SqlAlchemy.  

The current tablesorter looks to be from http://tablesorter.com -- it looks like it was last updated in 2008, written for jquery 1.2 and ships with 1.4.  It has been forked on github numerous times in the past 7 years for performance features and leveraging more modern jquery versions.   It just doesn't make sense using such an older jquery library that can create outdated version requirements.

A good candidate might be this one:  
   https://github.com/Mottie/tablesorter

This is a newer fork that has native theme for bootstrap support (and the toolbar uses bootstrap)

I think a correct "fix" to this issue would be:

1. Update the tablesorter to something modern (one of the newer/maintained forks, or another library)
2. Decide on a sorting mechanism, document it, and audit/apply to current panels as needed:
 a. Keep `pDebugSortable` (or rename)
 b. Use a CSS selector to target all tables in the tollbar's div.
 c. Some other approach

**Comment by stevepiercy** [comment]
tl;dr: I'm in favor of merging PR #239 now, and keeping this issue open until a better PR comes in.

I agree with @jvanasco's argument, provided there is a PR for any improvement over the original (and no longer updated) tablesorter plugin. Without a PR, I'm inclined to accept the technical debt with PR #239 because we've started the conversation about how to improve the situation (which reduces debt) instead of just commenting it without an explanation (which is the original event that brought about the debt in the first place).

**Comment by hamx0r** [comment]
I don't see real risk. Uncommenting it has been verified to enable the
feature we all want and appears to have no side effects. If people find
bugs, they can report them for fixing. If people want to replace with a
maintained table sorter, they can do that too. But withholding a feature
for fear of unknown impacts when we have several people using that feature
with success seems like excessive caution.

**Comment by jvanasco** [comment]
Uncommenting is a temporary fix for 2 panels.  This thread has had more interest in 1 day than in 3 years.  The risk is that this gets sidelined again.

Let's just channel this focus on a "forever fix".

What if we just: 

1. replace the tablesorter with the newer fork 'Mottie/tablesorter', that should work identical and be good for the next few jquery updates (i think v1.6 and prior is less backwards compatible) 
2. keep the `pDebugSortable` 
3. document the sortable css class under http://docs.pylonsproject.org/projects/pyramid-debugtoolbar/en/latest/#adding-custom-panels

If that sounds ok, I'll generate the PR.  I just want this fixed once and finally, because I would love to enable this on a suite of panels we maintain [and we run a heavily forked version of debugtoolbar, so merges are a pain :) ]

**Comment by stevepiercy** [comment]
I'm OK with deferring an immediate merge of PR #239 in favor of a proper fix as both @hamx0r and @jvanasco have mentioned in this discussion. @jvanasco thank you for taking on the task. I personally appreciate your diligence on the debug toolbar.

**Comment by jvanasco** [comment]
Looping back on this - both these PRs are bad:

* https://github.com/Pylons/pyramid_debugtoolbar/pull/239
* https://github.com/Pylons/pyramid_debugtoolbar/pull/243

Why?

After some more testing, I realized that both PRs will often disable the interactive-console on the exception view.  (oh no!)  Some browsers allow this to work, most seem to not.

The reason is that `exception.dbtmako` does not load the same javascript - not does it load js in the same way - as `toolbar.dbtmako`.  the call to `.tablesorting()` in `toolbar.js` will throw an error on the exception view, stall javacript, and prevent the debugger tools from being activated.  

at a minimum, wrapping the call in a try/catch works:

	try{
		// see http://mottie.github.io/tablesorter/docs/example-widget-bootstrap-theme.html
		$(".pDebugSortable").tablesorter({theme: "bootstrap",
										  widgets: ["uitheme", ],
										  headerTemplate: '{content} {icon}'
										  });
	}
	catch (err){
	}

there must be a better way to handle this.  my js skills are out of date, and I'm not entirely familiar with how `require.js` works.   standardizing the js/css on all these pages would probably fix this best.

**Comment by mmerickel** [comment]
closed via #256
